client/template.js

```diff
@@ -1,28 +1,67 @@
-module.exports = async(name, title = '', props = {})=>{
+const template = async function(name, title='', props = {}){
 	const HOMEBREWERY_PUBLIC_URL=props.config.publicUrl;
 
-	return `
-<!DOCTYPE html>
-<html>
-	<head>
-		<link href="//use.fontawesome.com/releases/v5.15.1/css/all.css" rel="stylesheet" />
-		<link href="//fonts.googleapis.com/css?family=Open+Sans:400,300,600,700" rel="stylesheet" type="text/css" />
-		<link href=${`/${name}/bundle.css`} rel='stylesheet' />
-		<link rel="icon" href="/assets/favicon.ico" type="image/x-icon" />
-		<meta property="og:title" content="${props.brew?.title || 'Homebrewery - Untitled Brew'}">
-		<meta property="og:url" content="${HOMEBREWERY_PUBLIC_URL}/${props.brew?.shareId ? `share/${props.brew.shareId}` : ''}">
-		<meta property="og:image" content="${props.brew?.thumbnail || `${HOMEBREWERY_PUBLIC_URL}/thumbnail.png`}">
-		<meta property="og:description" content="${props.brew?.description || 'No description.'}">
-		<meta property="og:site_name" content="The Homebrewery - Make your Homebrew content look legit!">
-		<meta property="og:type" content="article">
-		<meta name="twitter:card" content="summary">
-		<title>${title.length ? `${title} - The Homebrewery`: 'The Homebrewery - NaturalCrit'}</title>
-	</head>
-	<body>
-		<main id="reactRoot">${require(`../build/${name}/ssr.js`)(props)}</main>
-		<script src=${`/${name}/bundle.js`}></script>
-		<script>start_app(${JSON.stringify(props)})</script>
-	</body>
-</html>
-`;
+	const ogMeta = {
+		siteName    : 'The Homebrewery - Make your Homebrew content look legit!',
+		title       : 'The Homebrewery',
+		description : 'Homepage',
+		thumbnail   : `${HOMEBREWERY_PUBLIC_URL}/thumbnail.png`,
+		type        : 'website'
+	};
+
+	if(props.url.match(/\/share\/|\/edit\//)){
+		Object.assign(ogMeta, {
+			siteName    : null,
+			title       : props.brew.title || 'Homebrewery - Untitled Brew',
+			description : props.brew.description || 'No description.',
+			image       : props.brew.thumbnail || null,
+			type        : 'article'
+		});
+	} else if(props.url.match(/\/print\/|\/source\//)){
+		Object.assign(ogMeta, {
+			siteName    : null,
+			title       : props.brew.title ? `${props.brew.title} - ${props.url.match(/\/print\/|\/source\//)}` : 'Homebrewery - Untitled Brew',
+			description : props.brew.description || 'No description.',
+			image       : props.brew.thumbnail || null,
+			type        : 'article'
+		});
+	} else if(props.url.match(/\/user\//)){
+		Object.assign(ogMeta, {
+			siteName    : null,
+			title       : `${props.username} - The Homebrewery`,
+			description : `${props.username}'s user page.`,
+			image       : null,
+			type        : 'profile'
+		});
+	}
+
+	const ogTags = [];
+	Object.entries(ogMeta).forEach(([key, value])=>{
+		if(!value) return;
+		const tag = `<meta property="og:${key}" content="${value}">`;
+		ogTags.push(tag);
+	});
+	const ogMetaTags = ogTags.join('\n');
```

Comment:
I think the idea behind this PR is good, but I'm worried about adding yet another code location where each URL has different behavior. Could we possibly move the different metadata objects into `app.js`, and then in `template.js` the only change is this step you added that goes through a metadata object being passed in via `props`?

Reply:
Yes I can do that, that sounds much better. Even as I added this as-is I thought “this seems like a lot to put into template.js “

Reply:
Just to be sure I'm on the right track, is the `home page v3` comment on app.js wrong/a typo?  It should be `home page legacy`, right?   

```js
//Home page
app.get('/', (req, res, next)=>{
	req.brew = {
		text     : welcomeText,
		renderer : 'V3'
	};
	splitTextStyleAndMetadata(req.brew);
	return next();
});

//Home page v3
app.get('/legacy', (req, res, next)=>{
	req.brew = {
		text     : welcomeTextLegacy,
		renderer : 'legacy'
	};
	splitTextStyleAndMetadata(req.brew);
	return next();
});
```

Reply:
Just pushed a commit on this.  It's not complete, but I'm going to bed and thought I'd push this up in case someone can say "looks good so far" or "wait hold on a minute".  

If it looks okay, I do have a question-- how could I put metatags in the `/source/` page with this method?  It seems /source/ doesn't use the template, and has an empty `<head>` tag.

Reply:
I don't think source needs a meta tag, since it literally just sends the brew file contents. That is, if you visit the source link the *only* thing you should get is brew contents, free of any external tags or scripts.

Reply:
You don't think there is value in displaying an embed 'card' when posting a link to a /source/ page?  To me it just seems like any other page on the site.

But leaving the `head` empty is fine with me, too-- less work.

Reply:
> To me it just seems like any other page on the site.

In this case, it *isn't* a normal web page. It's meant to be just plain text, like opening a .txt file. That should be all you get.

Reply:
Okay, I just pushed a commit on this that I think is *close*, if not complete.  I've updated the original comment in this PR to reflect the current setup of the meta tags on each page, and there have been some changes from what I originally had.  

I guess there are two parts to review at this point:  the basic structure of my code, and the decisions on what each page should display.